library/core/src/str/validations.rs

```diff
@@ -112,172 +111,302 @@ where
+const STATE_MASK: u32 = (1 << BITS_PER_STATE) - 1;
+const STATE_CNT: usize = 9;
+#[allow(clippy::all)]
+const ST_ERROR: u32 = 0 * BITS_PER_STATE as u32;
+#[allow(clippy::all)]
+const ST_ACCEPT: u32 = 1 * BITS_PER_STATE as u32;
 
-/// Returns `true` if any byte in the word `x` is nonascii (>= 128).
+// After storing STATE_CNT * BITS_PER_STATE = 54bits on 64-bit platform, or (STATE_CNT - 5)
+// * BITS_PER_STATE = 24bits on 32-bit platform, we still have some high bits left.
+// They will never be used via state transition.
+// We merge lookup table from first byte -> UTF-8 length, to these highest bits.
+const UTF8_LEN_HIBITS: u32 = 4;
+
+static TRANS_TABLE: [u64; 256] = {
+    let mut table = [0u64; 256];
+    let mut b = 0;
+    while b < 256 {
+        // Target states indexed by starting states.
+        let mut to = [0u64; STATE_CNT];
+        to[0] = 0;
+        to[1] = match b {
+            0x00..=0x7F => 1,
+            0xC2..=0xDF => 2,
+            0xE0 => 3,
+            0xE1..=0xEC | 0xEE..=0xEF => 4,
+            0xED => 5,
+            0xF0 => 6,
+            0xF4 => 7,
+            0xF1..=0xF3 => 8,
+            _ => 0,
+        };
+        to[2] = match b {
+            0x80..=0xBF => 1,
+            _ => 0,
+        };
+        to[3] = match b {
+            0xA0..=0xBF => 2,
+            _ => 0,
+        };
+        to[4] = match b {
+            0x80..=0xBF => 2,
+            _ => 0,
+        };
+        to[5] = match b {
+            0x80..=0x9F => 2,
+            _ => 0,
+        };
+        to[6] = match b {
+            0x90..=0xBF => 4,
+            _ => 0,
+        };
+        to[7] = match b {
+            0x80..=0x8F => 4,
+            _ => 0,
+        };
+        to[8] = match b {
+            0x80..=0xBF => 4,
+            _ => 0,
+        };
+
+        // On platforms without 64-bit shift, align states 5..10 to 32-bit boundary.
+        // See docs above for details.
+        let need_align = cfg!(any(target_pointer_width = "16", target_pointer_width = "32"));
+        let mut bits = 0u64;
+        let mut j = 0;
+        while j < to.len() {
+            let to_off =
+                to[j] * BITS_PER_STATE as u64 + if need_align && to[j] >= 5 { 2 } else { 0 };
+            let off = j as u32 * BITS_PER_STATE + if need_align && j >= 5 { 2 } else { 0 };
+            bits |= to_off << off;
+            j += 1;
+        }
+
+        let utf8_len = match b {
+            0x00..=0x7F => 1,
+            0xC2..=0xDF => 2,
+            0xE0..=0xEF => 3,
+            0xF0..=0xF4 => 4,
+            _ => 0,
+        };
+        bits |= utf8_len << (64 - UTF8_LEN_HIBITS);
+
+        table[b] = bits;
+        b += 1;
+    }
+    table
+};
+
+#[cfg(not(any(target_pointer_width = "16", target_pointer_width = "32")))]
+#[inline(always)]
+const fn next_state(st: u32, byte: u8) -> u32 {
+    TRANS_TABLE[byte as usize].wrapping_shr(st as _) as _
+}
+
+#[cfg(any(target_pointer_width = "16", target_pointer_width = "32"))]
```

Comment:
Aside: there are a few targets that have 32 bit pointers but are capable of efficient 64 bit shifts. The most important once is wasm32, which has native 64 bit integer instructions and is often (but not always) executed on 64 bit capable hosts.

Reply:
Fixed.